### PR TITLE
Port the sns module to boto3

### PR DESCRIPTION
**Pull request by flowerysong**
##### SUMMARY

Ports the sns module to boto3 and AnsibleAWSModule. Adds support for protocol-specific messages targeting newer protocols, including 'lambda'.

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

sns

##### ANSIBLE VERSION
2.8

##### ADDITIONAL INFORMATION

```paste below
ok: [tasty-deepone.syslog.x.mail.umich.edu -> localhost] => {
    "changed": false, 
    "invocation": {
        "module_args": {
            "application": null, 
            "aws_access_key": null, 
            "aws_secret_key": null, 
            "ec2_url": null, 
            "email": null, 
            "email_json": null, 
            "http": null, 
            "https": null, 
            "lambda": "{\"source\": \"pandora.x.mail.umich.edu\", \"message\": \"H\\u00e9r fylgir stu\\u00f0ning hverri sannkenning, sv\\u00e1 sem k\\u00f6llu\\u00f0 er eggin \\u00f3h\\u00f6r\\u00f0, en fullhvatir mennirnir.\", \"sourcetype\": \"ansible\"}", 
            "message_attributes": null, 
            "message_structure": "json", 
            "msg": "Hér er stafasetning sú, er hætti ræðr ok kveðandi gerir.", 
            "profile": "vdc-nonprod", 
            "region": "us-west-2", 
            "security_token": null, 
            "sms": null, 
            "sqs": null, 
            "subject": "Hello from Ansible", 
            "topic": "rootmail", 
            "validate_certs": true
        }
    }, 
    "message_id": "d46335db-7f32-5843-a5f2-61cfbf9eb7b5", 
    "msg": "OK"
}
```

### CODE REVIEW

File: lib/ansible/modules/cloud/amazon/sns.py [hunk 1]
```diff
@@ -108,115 +110,116 @@
...
-            return topic[u'TopicArn']
-    return None
+
+    while True:
+        response = sns.list_topics(**sns_kwargs)
+        for topic in response['Topics']:
+            if topic['TopicArn'].endswith(lookup_topic):
+                return topic['TopicArn']
+
+        token = response.get('NextToken')
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/sns.py]: ListTopics has a [paginator](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/sns.html#SNS.Paginator.ListTopics)

File: lib/ansible/modules/cloud/amazon/sns.py [hunk 2]
```diff
@@ -108,115 +110,116 @@
...
+        subject=dict(type='str', default=None),
+        topic=dict(type='str', required=True),
+        application=dict(type='str', default=None),
+        message_attributes=dict(type='dict', default=None),
+        message_structure=dict(type='str', choices=['json', 'string'], default='json'),
     )
 
-    module = AnsibleModule(argument_spec=argument_spec)
+    for p in protocols:
+        argument_spec[p] = dict(type='str', default=None)
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/sns.py]: You can use just `dict()` rather than `dict(type='str', default=None)`

**Review by willthames — CHANGES_REQUESTED** [review]

**Comment by willthames** [comment]
This is great work, thanks. Could we have a test suite too please.

File: lib/ansible/modules/cloud/amazon/sns.py (hunk 1 above)
Comment (flowerysong) [reply lib/ansible/modules/cloud/amazon/sns.py]: Why, so it does.

File: lib/ansible/modules/cloud/amazon/sns.py [hunk 3]
```diff
@@ -108,115 +110,110 @@
...
-    result = response[u'ListTopicsResponse'][u'ListTopicsResult']
-    # topic names cannot have colons, so this captures the full topic name
+def arn_topic_lookup(client, short_topic):
     lookup_topic = ':{}'.format(short_topic)
-    for topic in result[u'Topics']:
-        if topic[u'TopicArn'].endswith(lookup_topic):
-            return topic[u'TopicArn']
+
+    paginator = client.get_paginator('list_topics')
+    topic_iterator = paginator.paginate()
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/sns.py]: Please wrap the client and paginator methods in a try/except and handle BotoCoreError and ClientError

**Review by s-hertel — COMMENTED** [review]
Looks good. Just could use some integration tests to help ensure backwards compatibility.

**Review by ryansb — APPROVED** [review]

**Comment by ryansb** [comment]
@willthames can you have a second look at this?

**Review by willthames — APPROVED** [review]

**Comment by willthames** [comment]
rebuild_merge

**Pull request merged**